### ISSUE TITLE
array.size overflows for large arrays on Windows

### ISSUE THREAD

**Issue by robintw** [dask/dask#1479]
Dask's implementation of `array.size` seems to overflow for moderately-large arrays on Windows:

```
x = da.ones((1491, 1162, 1240), chunks=(5, 1162, 1240))
x.size
# On OS X gives: 2148352080
# On Windows gives: -2146615216
```

This seems to be because of the default use of `np.int32` on Windows.

I came across this through xarray, and first reported it on the xarray mailing list, see https://groups.google.com/forum/#!topic/xarray/J5uaRsbkqIM. In that situation, I was actually getting an error from `sanitize_index` as called by `reshape` - but all of these were because of the overflow in `array.size`.

I've tried changing the implementation of `array.size` to add `dtype=np.int64`, but that seems to give a different error in `reshape` (`ValueError: number of elements must remain unchanged`) - so I assume the fix is a bit more complex than my naive fix...

**Comment by jcrist** [comment]
This should be fixed here: https://github.com/dask/dask/pull/1478. Can you give this fix a try?

**Comment by shoyer** [comment]
Looks like we also need to fix things on this line in reshape:
https://github.com/dask/dask/blob/master/dask/array/core.py#L2597

**Comment by jcrist** [comment]
Would you mind handling that in your reshape PR? Could also merge that PR, and just fix them here (possibly easier testing for @robintw).

**Comment by robintw** [comment]
Fixed by #1478.
